src/App.tsx

```diff
@@ -1,24 +1,155 @@
+    }
+
+    return () => {
+      clearTimeout(timeoutId);
+    };
+  }, [error]);
+
+  const activeTodos = todos.filter(todo => !todo.completed);
+  const completedTodos = todos.filter(todo => todo.completed);
+
+  const visibleTodos = useMemo(() => {
+    switch (filter) {
+      case FilterOptions.ACTIVE:
+        return activeTodos;
+
+      case FilterOptions.COMPLETED:
+        return completedTodos;
+
+      default:
+        return todos;
+    }
+  }, [todos, filter]);
+
+  const addTodo = async (title: string) => {
+    try {
+      const newTodo = {
+        title,
+        userId: USER_ID,
+        completed: false,
+      };
+
+      setTempTodo({
+        id: 0,
+        ...newTodo,
+      });
+
+      const createdTodo = await createTodo(newTodo);
+
+      setTodos(prevTodos => [...prevTodos, createdTodo]);
+    } catch {
+      setError('Unable to add todo');
+    } finally {
+      setTempTodo(null);
+    }
+  };
+
+  const deleteTodo = async (todoId: number) => {
+    try {
+      setLoadingTodos(prevIds => [...prevIds, todoId]);
+      await removeTodo(todoId);
+      setTodos(prevTodos => prevTodos.filter(todo => todo.id !== todoId));
+    } catch {
+      setError('Unable to delete a todo');
+    } finally {
+      setLoadingTodos(prevIds => prevIds.filter(id => id !== todoId));
+    }
+  };
+
+  const handleDeleteComplitedTodos = async () => {
+    const deletePromises = completedTodos.map(todo => deleteTodo(todo.id));
+
+    try {
+      await Promise.all(deletePromises);
+    } catch {
+      setError('Unable to delete todos');
+    }
+  };
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-loading-todos#react-todo-app-load-todos">React Todo App - Load Todos</a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <header className="todoapp__header">
+          <button
+            type="button"
+            className="todoapp__toggle-all active"
+            aria-label="active"
+          />
+          <TodoForm
+            setError={setError}
+            addTodo={addTodo}
+            tempTodo={tempTodo}
+          />
+        </header>
+
+        <TodoList
+          todos={visibleTodos}
+          tempTodo={tempTodo}
+          loadingTodos={loadingTodos}
+          deleteTodo={deleteTodo}
+        />
+
+        {todos.length > 0 && (
+          <footer className="todoapp__footer">
+            <span className="todo-count">
+              {`${activeTodos.length} items left`}
+            </span>
+
+            <Filter filter={filter} setFilter={setFilter} />
+
+            {completedTodos.length > 0 && (
```

Comment:
it's better to add `visibility: hidden` (add a proper class) instead of using conditional rendering to save the place of the button in the layout and keep the filter centered:
![image](https://github.com/mate-academy/react_todo-app-add-and-delete/assets/86923089/9752fcfb-8b5b-44bc-a4a2-13ef8783625e)

just FYI - implement it in the next step task :)